source/game/system/map/MapdataCheckPoint.cc

```diff
@@ -0,0 +1,258 @@
+#include "MapdataCheckPoint.hh"
+#include <Common.hh>
+#include <egg/math/Vector.hh>
+#include <game/system/CourseMap.hh>
+
+namespace System {
+
+MapdataCheckPoint::MapdataCheckPoint(const SData *data)
+    : m_rawData(data), m_nextCount(0), m_prevCount(0) {
+    u8 *unsafeData = reinterpret_cast<u8 *>(const_cast<SData *>(data));
+    EGG::RamStream stream = EGG::RamStream(unsafeData, sizeof(SData));
+    read(stream);
+    m_midpoint = EGG::Vector2f((left().x + right().x) / 2.0f, (left().y + right().y) / 2.0f);
+    m_dir = EGG::Vector2f(left().y - right().y, right().x - left().x);
+    m_dir.normalise();
+}
+
+void MapdataCheckPoint::read(EGG::Stream &stream) {
+    m_left.read(stream);
+    m_right.read(stream);
+    m_jugemIndex = stream.read_u8();
+    m_type = stream.read_s8();
+    m_prevPt = stream.read_u8();
+    m_nextPt = stream.read_u8();
+}
+
+/// @addr{0x80510D7C}
+/// @see MapdataCheckPoint::checkSectorAndCheckpointCompletion_
+MapdataCheckPoint::SectorOccupancy MapdataCheckPoint::checkSectorAndCheckpointCompletion(
```

Comment:
that makes sense; to differentiate it from the other function i went with `checkSectorsAndCompletions` (pluralizd) for now, since it calls the other function in a loop for all `m_nextPoints`